tests/e2e/k8s/util/kind.go

```diff
@@ -0,0 +1,425 @@
+}
+
+// Pod represents a kubernetes pod.
+type Pod struct {
+	// Name is the pod name.
+	Name string
+	// Namespace is the pod namespace.
+	Namespace string
+	// Image is the container image.
+	Image string
+	// Args are the container command line arguments.
+	Args []string
+}
+
+// KindCluster represents a kind kubernetes cluster.
+type KindCluster struct {
+	AsyncRunner
+
+	created   sync.WaitGroup
+	name      string
+	ip        string
+	cluster   *kind.Cluster
+	resources *resources.Resources
+	rawClient *http.Client
+}
+
+// Name returns the cluster name.
+func (c *KindCluster) Name() string {
+	return c.name
+}
+
+// IP returns the cluster IP.
+func (c *KindCluster) IP() string {
+	return c.ip
+}
+
+// Start the kind cluster.
+// Should be called only once.
+func (c *KindCluster) Start() {
+	c.created.Add(1)
+	c.Run(func() error {
+		// destroy cluster before creating (if exists)
+		_ = c.cluster.Destroy(context.Background())
+
+		// create cluster
+		_, err := c.cluster.Create(context.Background())
+		c.created.Done()
+		if err != nil {
+			return fmt.Errorf("unable to create kind cluster: %w", err)
+		}
+
+		// wait for controlplane to be ready and initialize clients
+		if err := c.initializeClients(); err != nil {
+			return err
+		}
+
+		// get cluster external-facing IP
+		if err := c.initClusterIP(); err != nil {
+			return fmt.Errorf("unable to get cluster IP: %w", err)
+		}
+
+		return nil
+	})
+}
+
+func (c *KindCluster) initializeClients() error {
+	client, err := klient.New(c.cluster.KubernetesRestConfig())
+	if err != nil {
+		return fmt.Errorf("error initializing API server client: %w", err)
+	}
+
+	if err := c.cluster.WaitForControlPlane(context.Background(), client); err != nil {
+		return fmt.Errorf("error waiting for controlplane to be ready: %w", err)
+	}
+
+	cfg := c.cluster.KubernetesRestConfig()
+	c.resources, err = resources.New(cfg)
+	if err != nil {
+		return fmt.Errorf("unable to initialize REST client: %w", err)
+	}
+
+	c.rawClient, err = rest.HTTPClientFor(cfg)
+	if err != nil {
+		return fmt.Errorf("unable to initialize raw REST HTTP client: %w", err)
+	}
+
+	return nil
+}
+
+func (c *KindCluster) initClusterIP() error {
+	var nodes v1.NodeList
+	if err := c.resources.List(context.Background(), &nodes); err != nil {
+		return fmt.Errorf("unable to list cluster nodes: %w", err)
+	}
+
+	if len(nodes.Items) != 1 {
+		return fmt.Errorf("expected a single node, but got: %v", nodes)
+	}
+
+	addresses := nodes.Items[0].Status.Addresses
+	for _, addr := range addresses {
+		if addr.Type == v1.NodeInternalIP {
+			if c.ip != "" {
+				return fmt.Errorf("expected a single node IP, but got: %v", addresses)
+			}
+
+			c.ip = addr.Address
+			if c.ip == "" {
+				return fmt.Errorf("got empty node IP address: %v", addr)
+			}
+		}
+	}
+
+	if c.ip == "" {
+		return fmt.Errorf("could not get node IP: %v", addresses)
+	}
+
+	return nil
+}
+
+// LoadImage loads a docker image to the cluster.
+// Assumes Start was already called.
```

Comment:
I don't think it's worth-while "protecting" the test code. (The next step would be writing a test for the test :) )
If some future dev will use this code and for some reason will decide to call `LoadImage` before `Start`, it will simply fail since the cluster does not exist.